Honor DOCKER_BUILDKIT in integration tests

tiborvass opened the pull request:
This PR refactors code in `integration/build/` and honors the DOCKER_BUILDKIT environment variable.

It introduces a new `integration/build/buildutil` package that could in the future be refactored into a easier to use docker client package.

All integration (non-cli) tests should pass with or without BuildKit enabled.

tonistiigi commented:
@tiborvass CI errors

thaJeztah commented on integration/build/util_test.go:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+func writeTarRecord(t *testing.T, w *tar.Writer, fn, contents string) {
```
not introduced in this PR, but just noticed this should probably have `t.Helper()`?

vdemeester commented on integration/build/util_test.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+func init() {
+	defaultBuildOptions = withBuilderVersion(types.ImageBuildOptions{
```
nit: why not

```go
var (
    defaultBuildOptions = withBuilderVersion(…)
)
```

vdemeester reviewed (approved):
LGTM with a small comment

tiborvass commented:
Actually I have found an issue with this PR when buildkit is tested.

Commit:
buildutil: new buildutil package to wrap client.Client.ImageBuild

This will allow to refactor the integration build tests to use structured
results instead of string comparisons.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
buildutil: support for session

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
integration: Honor DOCKER_BUILDKIT in integration tests

Signed-off-by: Tibor Vass <tibor@docker.com>

thaJeztah commented:
```
04:10:00 Running /go/src/github.com/docker/docker/integration/build/buildutil
04:10:00 env: './test.main': No such file or directory
```


Also some linting issues;

```
04:03:28 integration/build/buildutil/util.go:84:1:warning: exported function Build should have comment or be unexported (golint)
04:03:28 integration/build/buildutil/util.go:78:6:warning: exported type BuildInput should have comment or be unexported (golint)
04:03:28 integration/build/buildutil/util.go:51:1:warning: exported method BuildResult.CacheHit should have comment or be unexported (golint)
04:03:28 integration/build/buildutil/util.go:39:1:warning: exported method BuildResult.OutputContains should have comment or be unexported (golint)
04:03:28 integration/build/buildutil/util.go:30:6:warning: exported type BuildResult should have comment or be unexported (golint)
04:03:28 integration/build/buildutil/util.go:26:1:warning: exported function BuildKitEnabled should have comment or be unexported (golint)
```

tiborvass commented:
@thaJeztah fixed.

I moved `buildutil` to `builder/buildutil`.

Commit:
buildutil: Fix session logic with Dockerfile input

Signed-off-by: Tibor Vass <tibor@docker.com>

AkihiroSuda commented:
needs rebase

AkihiroSuda commented:
ping^^ @tiborvass

olljanat commented:
@tiborvass this one needs to be rebased

thaJeztah commented:
carried in https://github.com/moby/moby/pull/38415

olljanat commented:
Derek close
